### DISCUSSION TITLE
Bank File Format Discussion

### DISCUSSION THREAD

**Discussion by albedozero** [GeekFunkLabs/fluidpatcher#20] [General]
The version 0.5 release made some updates to the bank file format in order to work smoothly with new features. The new format is described in detail in [file_formats.md](https://github.com/albedozero/fluidpatcher/blob/master/patcher/file_formats.md). This means bank files created for older versions may need to be updated. This discussion gives tips for doing this. It is also a place for questions/feedback on bank files - if you have a question on how to do something with bank files, ask it here!

**Comment by albedozero** [comment]
Bank files created for version <0.5 will still work, but some keywords are deprecated and will be ignored.

- The `cc` keyword is replaced by `messages`, which can send any type of MIDI message. The keyword should contain a list of messages, formatted as `<type>:<channel>:<par1>:<par2>`.
- The `sysex` keyword is deprecated. Sysex messages can be sent using the `messages` keyword, formatted either as `sysex:<destination>:<filename.syx>` or `sysex:<destination>:<byte0>:<byte1>:<byte2>:...`
- the `effects` keyword is replaced by `ladspafx`, and its contents are formatted differently - see [file_formats.md](https://github.com/albedozero/fluidpatcher/blob/master/patcher/file_formats.md) for details.
- The `cclinks` keyword is deprecated. Instead, router rules can be used to route any type of MIDI message to change fluidsettings, control effects parameter ports, players, etc.

**Comment by yvmor** [comment]
Hello,
I discover fluidpatcher (fluidpatcher.pyw)  on a desktop computer under ubuntu OS
> You wrote "router rules can be used to route any type of MIDI message to change fluidsettings, control effects parameter ports"
So, could I transform a MIDI Control change and/or a cpress  to a pitch bend? 

> How to select audio alsa  or jack for fluidsynth ? (fluidpatcherconf.yaml mentions both !), I have no access to all the settings as I have when starting fluidsynth on a terminal.

> The connection between fluidsynth audio outputs and system audio is not automatic whereas the fluidsynth input is connected, I have to open a (new) terminal  and type: jack_connect fluidsynth:left system:plaback_1 

Thank you for any help
Anyway, the fluidpather application looks great !! and congratulations
All the best
Yves
r

**Reply by albedozero** [reply]
> You wrote "router rules can be used to route any type of MIDI message to change fluidsettings, control effects parameter ports"
> So, could I transform a MIDI Control change and/or a cpress  to a pitch bend?

Yes you can! Here's a rule that would do it:
```
router_rules:
- {type: cpress, type2: pbend, par1: 0-127=0-16383}
```
There's an example of pbend to cc in [bank0.yaml](https://github.com/albedozero/fluidpatcher/blob/2551f842201550fadf718224ac3d0dc02d3bbcd5/SquishBox/banks/bank0.yaml#L45) as well. [Lesson video](https://youtube.com/playlist?list=PL4a8Oe3qfS_-CefZFNYssT1kHdzEOdAlD) on this coming out (hopefully) soon.
 
> How to select audio alsa  or jack for fluidsynth ? (fluidpatcherconf.yaml mentions both !), I have no access to all the settings as I have when starting fluidsynth on a terminal.

FluidPatcher only reads `fluidsettings` on startup, so if you want to use the `fluidsettings-windows-lowlatency` or `fluidsettings-linux-jack` settings you can just copy-paste the values into the `fluidsettings` item.
 
> The connection between fluidsynth audio outputs and system audio is not automatic whereas the fluidsynth input is connected, I have to open a (new) terminal  and type: jack_connect fluidsynth:left system:plaback_1

You need `audio.jack.autoconnect: 1` - the above comment should fix it

**Reply by yvmor** [reply]
Thank you for the answers:
 I did not figure out it was so simple to transform cpress to pbend (which are not regular cc)
The given config file has three fluidsettings groups (windows, jack, alsa), is it the last one that is taken into account ?
I solve the audio connection problem, adding three lines in the reset function of class synth:
class synth
    connected = None
   ...
   def reset(self):
        if self.connected == None:
            os.popen('jack_connect fluidsynth:left system:playback_1').read()                 
            os.popen('jack_connect fluidsynth:right system:playback_2').read()
            self.connected = True
        FL.fluid_synth_system_reset(self.fsynth)

It is probably not a clean way to solve the problem, but it works in my config,
The fluidsetting line" audio.jack.autoconnect:1 " does not work whereas midi inputs ar automatically connected to fluidsynth

Thank you for your marvellous application
Yves

**Reply by albedozero** [reply]
It's only the group with the exact name `fluidsettings` that is read. `fluidsettings-linux-jack` and `fluidsettings-windows-wasapi` are ignored. I included them to show setups that can be used on specific systems. To use one of them, you can change its name to simply `fluidsettings` and delete the other groups.

Glad you like the software, hope you get some good use from it!
